x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -257,9 +257,9 @@ public IndexMetadata getSnapshotIndexMetaData(RepositoryData repositoryData, Sna
     }
 
     @Override
-    public void getRepositoryData(ActionListener<RepositoryData> listener) {
+    public void getRepositoryData(Executor responseExecutor, ActionListener<RepositoryData> listener) {
         try {
-            csDeduplicator.execute(listener.map(response -> {
+            csDeduplicator.execute(new ThreadedActionListener<>(responseExecutor, listener.map(response -> {
```

Comment:
I am trying to make sure this implementation adheres to the contract, i.e. responseExectuor is used if listener is _not_ completed in the calling thread. As is today, `csDeduplicator.execute` issues a network request and finishes on a transport worker thread. This means forking always happens. Even when calling thread is a transport thread (may not be possible), the forking still technically happens. Therefore, by the contract, we should always fork (except when executor is Direct) when completes the listener so that the usage of `ThreadedActionListener` is correct. Am I getting this right?

Reply:
Yes, that's right. The important property for this API is that the listener is either completed on the calling thread when `getRepositoryData` returns, or else it is completed on the given executor. We don't really mind which one happens, and it's ok to fork even if it technically wasn't necessary, although it'd be better to go down the fast path in the most common cases.